crates/compilers/src/preprocessor.rs

```diff
@@ -0,0 +1,643 @@
+use super::project::Preprocessor;
+use crate::{
+    flatten::{apply_updates, Updates},
+    multi::{MultiCompiler, MultiCompilerInput, MultiCompilerLanguage},
+    solc::{SolcCompiler, SolcVersionedInput},
+    Compiler, ProjectPathsConfig, Result, SolcError,
+};
+use alloy_primitives::hex;
+use foundry_compilers_artifacts::{
+    ast::SourceLocation,
+    output_selection::OutputSelection,
+    visitor::{Visitor, Walk},
+    ContractDefinitionPart, Expression, FunctionCall, FunctionKind, MemberAccess, NewExpression,
+    ParameterList, SolcLanguage, Source, SourceUnit, SourceUnitPart, Sources, TypeName,
+};
+use foundry_compilers_core::utils;
+use itertools::Itertools;
+use md5::Digest;
+use solar_parse::{
+    ast::{Span, Visibility},
+    interface::diagnostics::EmittedDiagnostics,
+};
+use std::{
+    collections::{BTreeMap, BTreeSet},
+    path::{Path, PathBuf},
+};
+
+/// Removes parts of the contract which do not alter its interface:
+///   - Internal functions
+///   - External functions bodies
+///
+/// Preserves all libraries and interfaces.
+pub(crate) fn interface_representation(
+    content: &str,
+    file: &PathBuf,
+) -> Result<String, EmittedDiagnostics> {
+    let mut spans_to_remove: Vec<Span> = Vec::new();
+    let sess =
+        solar_parse::interface::Session::builder().with_buffer_emitter(Default::default()).build();
+    sess.enter(|| {
+        let arena = solar_parse::ast::Arena::new();
+        let filename = solar_parse::interface::source_map::FileName::Real(file.to_path_buf());
+        let Ok(mut parser) =
+            solar_parse::Parser::from_source_code(&sess, &arena, filename, content.to_string())
+        else {
+            return;
+        };
+        let Ok(ast) = parser.parse_file().map_err(|e| e.emit()) else { return };
+        for item in ast.items {
+            if let solar_parse::ast::ItemKind::Contract(contract) = &item.kind {
```

Comment:
can use let Some else here too

Reply:
can you please detail how you see this changed

Reply:
doh, nm, got it, updated in https://github.com/foundry-rs/compilers/pull/252/commits/7d0edb851c0bf0234be4130e9950537187c38fab